lightning/src/ln/channelmanager.rs

```diff
@@ -3305,6 +3329,67 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			self.finish_force_close_channel(failure);
 		}
 	}
+
+	/// Gets a payment secret, payment hash, and absolute CLTV expiry for use in an invoice given
+	/// to a third party wishing to pay us.
+	///
+	/// A [`PaymentReceived`] event will only be generated if the payment secret matches a payment
+	/// secret fetched via this method or [`get_payment_secret`], and which is at least the `value`
+	/// provided here, if one is provided.
+	///
+	/// invoice_expiry_delta_blocks describes the number of blocks which the invoice is valid for,
+	/// in excess of the current block height. This should match the value set in the invoice.
+	///
+	/// May panic if invoice_expiry_delta_blocks is greater than one year of blocks (52,560).
+	pub fn get_payment_secret_preimage(&self, value_msat: Option<u64>, invoice_expiry_delta_blocks: u32) -> (PaymentHash, PaymentSecret) {
```

Comment:
This is roughly the top-level API I was picturing:
```
impl Invoice {
	pub fn from_channelmanager<Signer: Sign, M: Deref, T: Deref, K: Deref, F:
	Deref, L: Deref>(channelmanager: &ChannelManager<Signer, M, T, K, F, L>,
			our_node_privkey: SecretKey, amt_msat: u64, preimage: PaymentPreimage,
			payment_secret: PaymentSecret, description: String, network: Network) ->
		Result<Invoice, CreationError>
```
So I guess that makes me a bit confused where `get_payment_secret_preimage` and `get_payment_secret` fit into the equation. Maybe giving example usage in the docs would help?

FYI @jkczyz I think I told you I was creating this API^ but ended up passing the buck to Matt yesterday, who then created this PR.

Reply:
I think we should (a) drop the node_privkey parameter and expose it from ChannelManager, (b) make amount_msat Optional (so that people can do 0-value invoices), though I guess they could pass 0?, (d) drop the preimage and secret parameters.

Then that method would call `chanman.get_payment_secret_preimage(amount, DEFAULT_EXPIRY)` and use the payment secret and hash provided by that method, then calling `chanman.get_node_key` when signing.

Reply:
Ah okay... So when does the user have an opportunity to store the payment preimage/secret? (I know we're currently not dropping them on expiry, but I presume that we plan to.)

Reply:
Could I help by writing that utility based off this work? Or is some already working on it?

Reply:
Discussed offline, I have a half-finished utility so gonna update features and then PR it based on this PR.

Reply:
> So when does the user have an opportunity to store the payment preimage/secret?

I thought the goal was that they wouldn't have to? :) But, generally, if the user wants to store it, they can do it after calling `get_payment_secret_preimage` or, if they prefer to generate their own preimage, before calling `get_payment_secret`.